lib/db_connection.ex

```diff
@@ -1317,12 +1577,46 @@ defmodule DBConnection do
     end
   end
 
-  defp conclude(conn, result) do
+  defp conclude(%DBConnection{conn_ref: conn_ref} = conn, result) do
     case get_info(conn, nil) do
       {:ok, _conn_state, _meter} ->
-        {:ok, result}
+        result
       _ ->
-        {:error, :rollback}
+        throw({__MODULE__, conn_ref, :rollback})
+    end
+  end
+
+  defp reset(conn) do
+    case get_info(conn, nil) do
+      {:failed, conn_state, _meter} ->
+        put_info(conn, :ok, conn_state)
+      _ ->
+        :ok
+    end
+  end
+
+  defp transaction_error(conn, query, opts) do
+    run(conn, &run_transaction_error/5, query, meter(opts), opts)
+  end
+
+  defp run_transaction_error(conn, conn_state, query, meter, opts) do
+    meter = event(meter, query)
+    msg = "can not #{query} inside legacy transaction"
+    err = DBConnection.ConnectionError.exception(msg)
+    delete_disconnect(conn, conn_state, err, opts)
```

Comment:
We don't need but I worry what might happen if anyone hits this path. If we fail the transaction I guess that stop any issue - i.e. force a rollback at end of `transaction/3`.